xds/internal/xdsclient/creds/tls.go

```diff
@@ -0,0 +1,177 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package creds implements gRFC A65: mTLS Credentials in xDS Bootstrap File.
+package creds
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	_ "google.golang.org/grpc/credentials/tls/certprovider/pemfile" // for file_watcher provider
+)
+
+type tlsBundle struct {
+	jd                   json.RawMessage
+	transportCredentials credentials.TransportCredentials
+}
+
+// NewTLS returns a credentials.Bundle which delegates certificate loading to
+// a file_watcher provider for mTLS transport security. See gRFC A65.
+func NewTLS(jd json.RawMessage) (credentials.Bundle, error) {
+	cfg := &struct {
+		CertificateFile   string `json:"certificate_file"`
+		CACertificateFile string `json:"ca_certificate_file"`
+		PrivateKeyFile    string `json:"private_key_file"`
+	}{}
+
+	tlsConfig := tls.Config{}
+	if err := json.Unmarshal(jd, cfg); err != nil {
+		return nil, err
+	}
+
+	// We cannot simply always use a file_watcher provider because it behaves
+	// slightly differently from the xDS TLS config. Quoting A65:
+	//
+	// > The only difference between the file-watcher certificate provider
+	// > config and this one is that in the file-watcher certificate provider,
+	// > at least one of the "certificate_file" or "ca_certificate_file" fields
+	// > must be specified, whereas in this configuration, it is acceptable to
+	// > specify neither one.
+	//
+	// We only use a file_watcher provider if either one of them or both are
+	// specified.
+	if cfg.CACertificateFile != "" || cfg.CertificateFile != "" || cfg.PrivateKeyFile != "" {
+		// file_watcher currently ignores BuildOptions, but we set them for good
+		// measure.
+		opts := certprovider.BuildOptions{}
+		if cfg.CACertificateFile != "" {
+			opts.WantRoot = true
+		}
+		if cfg.CertificateFile != "" {
+			opts.WantIdentity = true
+		}
```

Comment:
This was a bit confusing for me, even more because I dont see other usages of BuildOptions where both `WantRoot` and `WantIdentity` are both set. 
Maybe let's just pass `nil` BuildOptions for now and leave a comment here mentioning that `file_watcher` ignores opts.

Reply:
I took your suggestion to not set the build options.

> This was a bit confusing for me, even more because I dont see other usages of BuildOptions where both WantRoot and WantIdentity are both set.

I'm not sure what bothers you with this, it seems legit for a provider to provide both a root cert and an identity, as file watcher does, no? That being said, I'm also unsure what those BuildOptions are really useful for: To me it would have been more natural to simply make that part of the config.

Reply:
I agree, there is a scope of minor cleanup in the `file_watcher` impl. Thanks!